### DISCUSSION TITLE
Can I track on Client that session just started or ended?

### DISCUSSION THREAD

**Discussion by yuzhakovvv** [nextauthjs/next-auth#1634] [Help] [answered]
Is there an event or callback on Client that helps to track that _session just started or ended_?

I didn't find anything related in the docs. In my case I need to do certain actions on Client after `signIn()` and `signOut()`. This is possible with `redirect: false` option for `credentials` and `email` providers. But for providers like `google` or `facebook` it becomes complicated.

**Comment by balazsorban44** [comment, marked as answer]
Have you seen the events callbacks?

https://next-auth.js.org/configuration/events

**Reply by yuzhakovvv** [reply]
@balazsorban44 these events run on the Server side

**Reply by balazsorban44** [reply]
Sorry, I looked over that you were looking for something in the frontend! Could you maybe give an example of what you are trying to do? Wouldn't `session` actually being available be a sign that you started a new session? 

`redirect: false` should work fine for `signOut` even for OAuth providers

**Reply by yuzhakovvv** [reply]
Example 1:
1. User opens an app for the first time. `session` is `undefined` and loading
2. `session` becomes `null` which means that user isn't signed in
3. User signs in with `signIn()` and `google` provider
4. User is redirected to Google Auth page and enters his credentials
5. User is redirected back to an app. `session` is `undefined` and loading
6. `session` becomes an `object`
7. _My logic should start here as new session started_

Transition of `session` from `undefined` to  an `object` is not a way to identify that session just started. Here is why:

Example 2:
1. User is already signed in. `session` is  an `object`
2. User reloads an app.`session` is `undefined` and loading
3. `session` becomes  an `object`
4. _My logic should NOT start here as new session didn't start_

Here we also have a transition from `undefined` to  an `object` and now session isn't just started.

**Reply by yuzhakovvv** [reply]
`session` transition from `null` to an `object` can be an identifier that session started. But only for providers that can set `redirect: false` and avoid page reload. Because without page reload it's possible to catch this transition.

**Reply by balazsorban44** [reply]
I am still not sure what kind of logic you are trying to run, but maybe localStorage or sessionStorage could help?

**Reply by yuzhakovvv** [reply]
In our case after Sign In we do:
1. Track of analytics "Signed In" client event
2. Redirect to a route that can happen only _after_ Sign In and getting user data. Can't use `callbackUrl` here
3. Manually create a cookie 

@balazsorban44 so as I understand there is no way to track on the Client that user _just Signed In_ in the current version of `next-auth`?

**Reply by balazsorban44** [reply]
I think you are trying to tie the `signIn` client-side function call with the `useSession()` hook across page refreshes. React state is blown away on page refresh and so in your comment https://github.com/nextauthjs/next-auth/discussions/1634#discussioncomment-562798 in Example 1, I think 1. and 5. are indistinguishable without some extra information saved on the user's machine first (like localStorage, sessionStorage, or a client-side cookie I guess(?)).

If I understand your requirements, I don't think currently we have anything built-in for this, but you could probably solve it with some kind of a persistent client-side mechanism.

1. User opens an app for the first time. `session` is undefined and loading
2. session becomes `null` which means that user isn't signed in. Save this information into some persistent client cache
3. User signs in with `signIn()` and google provider
4. User is redirected to Google Auth page and enters his credentials
5. User is redirected back to an app. `session` is `undefined` and loading
6. `session` becomes an object
7. If client cache signals it's the first sign-in, run my logic and remove the information from the cache
8. Page refresh wouldn't run 7., as the cache doesn't signal first sign-in anymore

Maybe @iaincollins have a better idea.

**Reply by yuzhakovvv** [reply]
This is exactly what I'm doing now. Using a `?signin=true` query param as persistent client state as others (localStorage, sessionStorage, or a client-side cookie) may be affected by user's browser settings.

I was wondering if there is a simpler solution out of the box 🤓

**Reply by balazsorban44** [reply]
I don't THINK there is currently, but I may be wrong. We are open for suggestions though, and if you have a good idea please open a PR! 😊
